src/core/cluster.py

```diff
@@ -0,0 +1,211 @@
+import logging
+from typing import TYPE_CHECKING, Literal, Set
+
+from ops.model import Relation
+from src.core.models import ZKClient, ZKCluster, ZKServer
+
+if TYPE_CHECKING:
+    from charm import ZooKeeperCharm
+
+
+logger = logging.getLogger(__name__)
+
+# TODO: move to some better literals place
+SUBSTRATE = Literal["vm", "k8s"]
+PEER = "cluster"
+REL_NAME = "zookeeper"
+
+
+class ClusterState:
+    def __init__(self, charm: "ZooKeeperCharm", substrate: SUBSTRATE):
+        self.charm = charm
+        self.substrate: SUBSTRATE = substrate
+
+    # --- RAW RELATION ---
+
+    @property
+    def peer_relation(self) -> Relation | None:
+        return self.charm.model.get_relation(PEER)
+
+    @property
+    def client_relations(self) -> Set[Relation]:
+        return set(self.charm.model.relations[REL_NAME])
+
```

Comment:
I think this is really useful. We do _a lot_ of checks to see if the relation even exists. So having that access done through the centralized state will help to do both relation checks, and data access in the same place.